cmd/cortex/main.go

```diff
@@ -139,21 +151,34 @@ func main() {
 	switch cfg.mode {
 	case modeDistributor:
 		cfg.distributorConfig.Ring = r
-		cfg.distributorConfig.ClientFactory = func(address string) (*distributor.IngesterClient, error) {
-			return distributor.NewIngesterClient(address, cfg.remoteTimeout)
-		}
 		setupDistributor(cfg.distributorConfig, chunkStore, router.PathPrefix("/api/prom").Subrouter())
 
 	case modeIngester:
 		cfg.ingesterConfig.Ring = r
-		registration, err := ring.RegisterIngester(consul, cfg.listenPort, cfg.numTokens)
+		registration, err := ring.RegisterIngester(consul, cfg.listenPort, cfg.ingesterConfig.GRPCListenPort, cfg.numTokens)
 		if err != nil {
 			// This only happens for errors in configuration & set-up, not for
 			// network errors.
 			log.Fatalf("Could not register ingester: %v", err)
 		}
 		ing := setupIngester(chunkStore, cfg.ingesterConfig, router)
 
+		// Setup gRPC server
+		lis, err := net.Listen("tcp", fmt.Sprintf(":%d", cfg.ingesterConfig.GRPCListenPort))
+		if err != nil {
+			log.Fatalf("failed to listen: %v", err)
+		}
+		grpcServer := grpc.NewServer(
+			grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(
+				cortex_grpc_middleware.ServerInstrumentInterceptor(requestDuration),
+				cortex_grpc_middleware.ServerLoggingInterceptor(cfg.logSuccess),
+				cortex_grpc_middleware.ServerUserHeaderInterceptor,
+			)),
+		)
+		cortex.RegisterIngesterServer(grpcServer, ing)
+		go grpcServer.Serve(lis)
+		defer grpcServer.Stop()
```

Comment:
Yeah.  I think we'll end up breaking this into a bunch of binaries / images next.